services/gateway/endpoints/src/main/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsRoute.java

```diff
@@ -0,0 +1,296 @@
+                    .trace("Ditto: {}", dittoHeaders);
+        }
+
+        // create a reader for the message
+        final MessageReader reader = HttpMessageFactory.createReader(acceptor -> {
+
+            // NOTE: this acceptor may be run multiple times by the message reader
+
+            // record if we saw the content type header
+            final AtomicBoolean sawContentType = new AtomicBoolean();
+            // consume the HTTP request headers
+            ctx.getRequest().getHeaders().forEach(header -> {
+                if (header.lowercaseName().equals(DittoHeaderDefinition.CONTENT_TYPE.getKey())) {
+                    sawContentType.set(true);
+                }
+                acceptor.accept(header.name(), header.value());
+            });
+
+            if (!sawContentType.get()) {
+                // we didn't see the content type in the header, so extract it from akka's request
+                acceptor.accept(DittoHeaderDefinition.CONTENT_TYPE.getKey(), ctx.getRequest().entity().getContentType().mediaType().toString());
+            }
+        }, payload.toArray());
+
+        return reader.toEvent();
+    }
+
+    private CloudEvent validateCloudEvent(final CloudEvent cloudEvent, final RequestContext ctx, final DittoHeaders dittoHeaders) {
+
+        if (cloudEvent.getData() == null) {
+            throw MissingPayloadException
+                    .withDetailedInformationBuilder()
+                    .dittoHeaders(dittoHeaders)
+                    .build();
+        }
+
+        LOGGER
+                .withCorrelationId(dittoHeaders)
+                .debug("Cloud event: {}", cloudEvent);
+
+        ensureDataContentType(cloudEvent.getDataContentType(), ctx, dittoHeaders);
+        ensureDataSchema(cloudEvent.getDataSchema(), ctx, dittoHeaders);
+
+        return cloudEvent;
+    }
+
+    private Optional<Signal<?>> jsonToDittoSignal(@Nullable final CloudEventData data, final DittoHeaders dittoHeaders) {
+        if (data == null) {
+            return Optional.empty();
+        }
+        final byte[] payload = data.toBytes();
+        if (payload == null || payload.length == 0) {
+            return Optional.empty();
+        }
+
+        final JsonObject jsonObject = JsonObject.of(payload);
+        LOGGER
+                .withCorrelationId(dittoHeaders)
+                .debug("JSON: {}", jsonObject);
+
+        final JsonifiableAdaptable jsonifiableAdaptable = ProtocolFactory.jsonifiableAdaptableFromJson(jsonObject);
+        return Optional.of(PROTOCOL_ADAPTER
+                .fromAdaptable(jsonifiableAdaptable)
+                .setDittoHeaders(dittoHeaders));
+    }
+
+    private void ensureDataContentType(@Nullable final String dataContentType,
+                                       final RequestContext ctx,
+                                       final DittoHeaders dittoHeaders) {
+
+        if (dataContentType == null || !mediaTypeJsonWithFallbacks.contains(dataContentType)) {
```

Comment:
I think that this "contains" check is problematic as the `dataContentType` could also contain a charset.
The javadoc of `getDataContenType()` says that it MUST follow RFC2046 which can contain the optional charset.

Reply:
Yes, but I think Ditto can define what it expects in there. Same a the standard HTTP `Content-Type` header … so requiring `application/json` from a Ditto side is fine I think.

Reply:
Hm, some tools may always include the charset even if not explicitly specified, so I think that this would also be true for backend applications emitting cloud-events.

I could also work on this part (also supporting the Ditto Protocol content-type) if you like.